Source/Initialization/InjectorMomentum.H

```diff
@@ -220,6 +220,49 @@ private:
     int m_flux_direction;
 };
 
+
+// struct whose getMomentum returns momentum for 1 particle, from random
+// waterbag distribution (i.e. a uniform distribution u_min < u < u_max)
+struct InjectorMomentumWaterbag
+{
+    InjectorMomentumWaterbag (amrex::Real a_ux_min, amrex::Real a_uy_min,
+                              amrex::Real a_uz_min, amrex::Real a_ux_max,
+                              amrex::Real a_uy_max, amrex::Real a_uz_max) noexcept
+        : m_ux_min(a_ux_min), m_uy_min(a_uy_min), m_uz_min(a_uz_min),
+          m_ux_max(a_ux_max), m_uy_max(a_uy_max), m_uz_max(a_uz_max)
+        {
+            m_Dux = m_ux_max - m_ux_min;
+            m_Duy = m_uy_max - m_uy_min;
+            m_Duz = m_uz_max - m_uz_min;
+            m_ux_h = 0.5 * (m_ux_max - m_ux_min);
+            m_uy_h = 0.5 * (m_uy_max - m_uy_min);
+            m_uz_h = 0.5 * (m_uz_max - m_uz_min);
+        }
+
+    AMREX_GPU_HOST_DEVICE
+    amrex::XDim3
+    getMomentum (amrex::Real /*x*/, amrex::Real /*y*/, amrex::Real /*z*/,
+                 amrex::RandomEngine const& engine) const noexcept
+    {
+        return amrex::XDim3{m_ux_min + amrex::Random(engine) * m_Dux,
```

Comment:
This generates a cube in velocity space and so will have a correlation along the coordinate axes. Would a sphere in velocity space be better since it would not have that correlation?

Reply:
I think you raise a really interesting point and I'm not sure I get the full picture on when this would be an actual issue or not. The classical waterbag is spherical in the momenta, so maybe the naming I'm using is inappropriate, but spherical momentum distributions would constrain all the momentum direction to the same maximum value (i.e. random direction for a given kinetic energy), which sometimes might be unnecessary if I already know I'm interested in a subset of those particles. Elliptical, perhaps? I wonder if all of those possibilities serve for different purposes (and at that point it might be worth having all of them available). In fact, it probably depends on the application and often one would have different constraints, for example when generating a beam you select the energy (with energy spread), main direction and RMS divergence.

Reply:
A general approach to the classical waterbag allows for a distribution on a phase space of any dimension (2D, 4D, or 6D), with an arbitrary covariance matrix.  The idea is to uniformly populate a phase space ball of the appropriate dimension, and then to apply a linear coordinate transformation to produce the correct covariance matrix.  The 6D version is used often in Impact-T.  In 2D, this would produce a uniformly-filled ellipse.

Reply:
The details of how we do this in the 6D case in ImpactX can be found here:
[](https://github.com/ECP-WarpX/impactx/blob/development/src/particles/distribution/Waterbag.H)

Reply:
@titoiride do you potentially have time to update this? :)

Reply:
Thanks for the feedback, I'll resume working on this PR ASAP.

Reply:
> A general approach to the classical waterbag allows for a distribution on a phase space of any dimension (2D, 4D, or 6D), with an arbitrary covariance matrix. The idea is to uniformly populate a phase space ball of the appropriate dimension, and then to apply a linear coordinate transformation to produce the correct covariance matrix. The 6D version is used often in Impact-T. In 2D, this would produce a uniformly-filled ellipse.

The "design" problem I see here is that the struct we're introducing here is an `InjectorMomentum`. So, it only outputs a momentum vector as in WarpX the coordinates of the particle distribution are specified using other parameters, so it's impractical to introduce correlations. Generating momenta in a uniform sphere on the other hand is easy.

Reply:
Understood, thanks.  In that case, I don't see a direct analogy with anything in Impact-T, but I agree with dpgrote that having a distribution that is isotropic in momentum space would be more relevant physically.  One option could be to start with a uniformly-populated sphere in momentum space, and then apply a linear transformation to allow user-specified correlations between px,py,pz if you want them (ie, an ellipsoid).